### ISSUE TITLE
Error Editing Deleted Instance

### ISSUE THREAD

**Issue by RobertBColton** [IsmAvatar/LateralGM#482]
This is another undo bug that is a regression and first appears in 1.8.5 but not 1.8.4 and earlier. If you create a new project, new object, new room. Open the room and place an instance of the object. Right click or hit undo to delete the object. Then focus one of the instance property fields. You will be met with an exception.

```
Stack trace:
java.lang.NullPointerException
	at org.lateralgm.subframes.RoomFrame.focusGained(RoomFrame.java:2372)
	at java.desktop/java.awt.AWTEventMulticaster.focusGained(AWTEventMulticaster.java:228)
	at java.desktop/java.awt.Component.processFocusEvent(Component.java:6525)
	at java.desktop/javax.swing.JFormattedTextField.processFocusEvent(JFormattedTextField.java:621)
	at org.lateralgm.components.NumberField.processFocusEvent(NumberField.java:150)
	at java.desktop/java.awt.Component.processEvent(Component.java:6392)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5012)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4844)
	at java.desktop/java.awt.KeyboardFocusManager.redispatchEvent(KeyboardFocusManager.java:1950)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.typeAheadAssertions(DefaultKeyboardFocusManager.java:1050)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:674)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:4893)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4844)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:391)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:391)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

**Comment by RobertBColton** [comment]
Alright, I got to the actual bottom of this issue. The reason the selected index of the list is not being updated to -1, causing the exception, when a selected instance is removed has to do with mistakes that were made in the underlying model classes.

The object instance list in the room editor uses `ArrayListModel`:
https://github.com/IsmAvatar/LateralGM/blob/bc4ba651cb78e1a7b930b13a460deea5154560fc/org/lateralgm/subframes/RoomFrame.java#L873
`ArrayListModel` listens for changes to its underlying `ActiveArrayList`:
https://github.com/IsmAvatar/LateralGM/blob/8bbd9c76ce77a50919a5db1601a9e44e1c6ead15/org/lateralgm/ui/swing/util/ArrayListModel.java#L30
`ActiveArrayList` fires a list update event when an item is added or removed:
https://github.com/IsmAvatar/LateralGM/blob/8bbd9c76ce77a50919a5db1601a9e44e1c6ead15/org/lateralgm/util/ActiveArrayList.java#L82
`ArrayListModel` will then call `contentsChanged()` instead of `intervalAdded` or `intervalRemoved`:
https://github.com/IsmAvatar/LateralGM/blob/8bbd9c76ce77a50919a5db1601a9e44e1c6ead15/org/lateralgm/ui/swing/util/ArrayListModel.java#L87

The problem is that `contentsChanged()` is not correct for adding or removing items. The documentation is actually pretty clear about this, so I'll let it explain.
> For example, this is sent when an item has been replaced.
https://docs.oracle.com/javase/7/docs/api/javax/swing/event/ListDataListener.html#contentsChanged(javax.swing.event.ListDataEvent)

**Comment by RobertBColton** [comment]
Closing as resolved by #484 which fixes egofree's check to see if the selected instance is valid.
